qiskit/circuit/quantumcircuit.py

```diff
@@ -5936,6 +5931,14 @@ def append(self, instruction):
         # QuantumCircuit._append is semi-public, so we just call back to it.
         return self.circuit._append(instruction)
 
+    def extend(self, data: CircuitData):
+        def update_param(op):
+            if isinstance(op, Instruction):
+                self.circuit._update_parameter_table(op)
+
+        self.circuit._data.extend(data)
+        data.foreach_op(update_param)
+
```

Comment:
This feels a bit prone to getting out-of-sync with `QuantumCircuit._append` here (it's already not identical).  It feels like the main performance reason for this `extend` method to exist is more efficient space reservation in `CircuitData`; perhaps it would be best to just reserve `len(data)` spaces, then call `Scope.append` in a loop to ensure everything stays in sync?

Reply:
It's a bit more than the preallocation in this case. Under the hood, `CircuitData::extend` also has special casing for when the source data is a `CircuitData` instance, in which case it skips the creation and destruction of ephemeral `CircuitInstruction` instances during the copy. In my testing, that seemed to help with performance, though I don't have an exact comparison handy.

To be symmetrical with the relationship between `QuantumCircuit._append` and `_OuterCircuitScopeInterface.append`, we could introduce a new `QuantumCircuit._extend` method and call it from `_OuterCircuitScopeInterface.extend`. I didn't opt for that initially since `QuantumCircuit._append` is sort of part of the public API (I think), whereas a `QuantumCircuit._extend` certainly wouldn't be since it operates on the private `CircuitData` type.

Reply:
I'd definitely not want to make a split `extend`/`_extend` - that would (if anything) exacerbate the problem of having the append logic in several places.  My worry is that this is a second place where any tracking the inner `_append` primitive does needs to be synced with, whereas previously it was just one location, and this change is already introducing an inconsistency - `_append` sets `self.duration = None` and `self.unit = "dt"` on all appends, while this doesn't.  I don't know exactly _why_ `_append` does that, but it worries me that we're introducing an inconsistency (well, more replacing `compose`'s blatant disregard for the data model with something that appears more inline but isn't entirely).

In a compiled language we'd obviously factor out the checks into an inlined function, but in Python we'll have a little bit of overhead.  I just tried it with a helper function in a `qiskit.qasm2.loads`-based workflow, which typically has a large dependence on the speed of `_append`, and it wasn't noticeable, though, so maybe doing this works:

```python
class QuantumCircuit:
    def _append(self, instruction):
        # ... whatever setup there is ...
        self._data.append(instruction)
        self._track_operation(instruction.operation)
        return whatever_it_is

    def _track_operation(self, operation: Operation):
        """Sync all non-data-list internal data structures for a newly tracked operation."""
        if isinstance(operation, Instruction):
            self._update_parameter_table(operation)
        self.duration = None
        self.unit = "dt"

class OuterCircuitScope(ScopeInterface):
    def extend(self, data):
        self.circuit._data.extend(data)
        data.foreach_op(self.circuit._track_operation)
```

Reply:
That seems like a great idea to me. Done in f27d2ca01f00db8b0c2dcc49d58a7c961aaa1f5c.